File: internal/connector/tegrauefi/verify.go
package tegrauefi

// Boot health and platform-update verification: ports of
// verify-bootloader-update (the cascade) and abort-blupdate, plus the
// RootfsStatusSlot health check used by the boot verify unit.

import (
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// ESRT entry0 last_attempt_status codes (validated: t234 incident analysis
// + t264 Phase 1). 0 = success; 1-6 = standard UEFI capsule errors; the
// NVIDIA-specific codes and vendor range are documented per NVIDIA L4T.
const (
	esrtSuccess   = 0
	esrtUEFIErrLo = 1 // 1..6: standard UEFI capsule errors
	esrtUEFIErrHi = 6
	// 6163 = LAS_ERROR_BOOT_CHAIN_UPDATE_CANCELED (NOT "CheckTheImage failed"; that
	// was a mis-read of a different firmware's code range). Proven from the r39.2
	// edk2-nvidia source: EDK2 FmpDevicePkg base LAST_ATTEMPT_STATUS_DEVICE_LIBRARY_MIN
	// = 0x1800 (6144), and the enum entry in TegraFmp.c is #19 → 6163. Set by
	// FmpTegraCheckImage (TegraFmp.c) via BootChainDxe.c BootChainCheckAndCancelUpdate,
	// which CANCELS the capsule while BootChainFwNext OR BootChainFwStatus exists (an
	// "FMP conflict" with a pending nvbootctrl-style FW-chain switch). SwapSlot now
	// clears both before staging (settleBootChain), so a fresh 6163 here means the
	// chain re-entered a pending state after staging — a real defect, not stale cruft.
	esrtBootChainUpdateCanceled = 6163
	esrtNvidiaSKUMismatch       = 6164   // device SKU not in the capsule's BUP
	esrtNvidiaVendorLo          = 0x1000 // 0x1000..0x4000: NVIDIA vendor error range
	esrtNvidiaVendorHi          = 0x4000
)

// BootIsCompromised reports whether the firmware flagged the slot we
// actually booted (RootfsStatusSlot status != 0) — a 0xFF right after a
// swap means UEFI burned the retry budget. The engine calls this only
// while an update is pending.
//
// Only the BOOTED slot is evidence about this boot's health. The previous
// implementation scanned both slots and flagged failure if EITHER was
// non-normal, so a stale 0xFF left on the inactive/old slot false-positived
// a perfectly healthy update (WDY-1742). Firmware fallback — running a
// different slot than the update targeted — is detected separately by the
// engine's running-slot vs target-slot check, not here.
//
// Conservative on uncertainty: if we cannot determine the current slot, or
// the booted slot's status var is absent or not the validated 8-byte layout,
// we return "not compromised" rather than crying wolf. The 8-byte format is
// confirmed on t234 (incl. Orin Nano r39.2, read off the device) and t264;
// an unexpected size (the other WDY-1742 failure mode) is treated as
// inconclusive instead of forcing a rollback.
// The engine's slot check and the ESRT platform-verify cascade remain the
// authoritative guards, so this loses no genuine-fallback detection.
func (c *Controller) BootIsCompromised() (bool, error) {
	cur, err := c.CurrentSlot()
	if err != nil {
		slog.Warn("boot health: cannot determine current slot; skipping efivar check", "err", err)
		return false, nil
	}

	raw, err := readStatus(c.statusVar(cur))
	if os.IsNotExist(err) {
		return false, nil // no status var for the booted slot: nothing to flag
	}
	if err != nil {
		return false, fmt.Errorf("boot health: %w", err)
	}
	if !statusIsWellFormed(raw) {
		slog.Warn("boot health: RootfsStatusSlot has unvalidated format; treating as inconclusive",
			"slot", cur.String(), "bytes", len(raw))
		return false, nil
	}
	if !statusIsNormal(raw) {
		slog.Warn("boot health: firmware flagged the booted slot unhealthy", "slot", cur.String())
		return true, nil
	}
	return false, nil
}

// VerifyPlatformUpdate is the verify-bootloader-update cascade. The
// running rootfs's marker file is the source of truth for whether a
// bootloader update was part of this deployment (same rule as staging);
// blUpdate from the manifest is informational.
//
//  1. PRIMARY:   bootloader version changed vs the value saved at swap
//  2. SECONDARY: ESRT last_attempt_status == 0
//  3. FALLBACK:  we booted, assume success — but say so loudly
//
// Validated ESRT codes (t234 incident analysis + t264 Phase 1):
// 0 success; 1-6 standard UEFI capsule errors; 6163 = LAS_ERROR_BOOT_CHAIN_UPDATE_CANCELED
// (capsule canceled by a pending FW-chain switch — see the const above);
// 6164 NVIDIA SKU mismatch; 0x1000-0x4000
// NVIDIA vendor range. nvbootctrl's own capsule status is NOT consulted —
// NVIDIA documents it as unreliable.
func (c *Controller) VerifyPlatformUpdate(blUpdate bool) error {
	if _, err := os.Stat(c.RootDir + MarkerPath); err != nil {
		if blUpdate {
			slog.Info("platform verify: manifest declared bootloader_update but the running rootfs has no marker; skipping")
		}
		return nil
	}
	slog.Info("platform verify: checking bootloader update (version + ESRT cascade)")

	// 1) version comparison
	if before, err := os.ReadFile(c.blVersionBeforePath()); err == nil {
		after, verr := c.bootloaderVersion()
		if verr == nil {
			if strings.TrimSpace(string(before)) != after {
				_ = os.Remove(c.blVersionBeforePath())
				slog.Info("platform verify: bootloader version changed — capsule applied", "version", after)
				return nil // version changed: capsule applied
			}
			slog.Info("platform verify: bootloader version unchanged; checking ESRT", "version", after)
		} else {
			slog.Warn("platform verify: could not read bootloader version; checking ESRT", "err", verr)
		}
	} else {
		slog.Warn("platform verify: no pre-update bootloader version recorded; checking ESRT")
	}

	// 2) ESRT verdict
	if raw, err := os.ReadFile(c.RootDir + ESRTStatusPath); err == nil {
		status, perr := strconv.Atoi(strings.TrimSpace(string(raw)))
		if perr != nil {
			return fmt.Errorf("platform verify: unparseable ESRT status %q", strings.TrimSpace(string(raw)))
		}
		switch {
		case status == esrtSuccess:
			_ = os.Remove(c.blVersionBeforePath())
			slog.Info("platform verify: ESRT reports success", "status", esrtSuccess)
			return nil
		case status >= esrtUEFIErrLo && status <= esrtUEFIErrHi:
			return fmt.Errorf("platform verify: ESRT status %d (standard UEFI capsule error)", status)
		case status == esrtBootChainUpdateCanceled:
			return fmt.Errorf("platform verify: ESRT status %d (LAS_ERROR_BOOT_CHAIN_UPDATE_CANCELED — capsule canceled by a pending FW-chain switch despite settleBootChain; the chain re-entered a pending state after staging)", esrtBootChainUpdateCanceled)
		case status == esrtNvidiaSKUMismatch:
			return fmt.Errorf("platform verify: ESRT status %d (NVIDIA: device SKU not included in the capsule's BUP)", esrtNvidiaSKUMismatch)
		case status >= esrtNvidiaVendorLo && status <= esrtNvidiaVendorHi:
			return fmt.Errorf("platform verify: ESRT status %d (NVIDIA vendor error)", status)
		default:
			slog.Warn("platform verify: unknown ESRT status; falling back to boot-success", "status", status)
		}
	} else {
		slog.Warn("platform verify: ESRT not readable; falling back to boot-success")
	}

	// 3) fallback: the system booted to this point
	slog.Warn("platform verify: could not confirm the bootloader update via version or ESRT; the system booted, assuming success — manual check recommended")
	_ = os.Remove(c.blVersionBeforePath())
	return nil
}

// AbortPlatformUpdate unstages a capsule that has not been processed:
// removes TEGRA_BL.Cap from the ESP (port of abort-blupdate) and disarms
// the OsIndications capsule bit so the firmware will not look for one.
// No-op when nothing is staged.
func (c *Controller) AbortPlatformUpdate() error {
	staged := false

	if espDir, err := c.espMountpoint(); err == nil {
		cap := filepath.Join(espDir, ESPCapsuleRel)
		if _, err := os.Stat(cap); err == nil {
			staged = true
			if err := os.Remove(cap); err != nil {
				return fmt.Errorf("abort platform update: %w", err)
			}
		}
	}

	osiPath := filepath.Join(c.EfivarsDir, "OsIndications-"+EfiGlobalGUID)
	if err := clearOsIndicationsCapsuleBit(osiPath); err != nil {
		return fmt.Errorf("abort platform update: %w", err)
	}

	if staged {
		_ = os.Remove(c.blVersionBeforePath())
	}
	return nil
}

File: swift/Sources/TegraUEFI/Verify.swift
import Connector
import PlatformIO

// Port of `internal/connector/tegrauefi/verify.go`'s `VerifyPlatformUpdate`
// (the bootloader-version + ESRT cascade) and `AbortPlatformUpdate`
// (unstage a not-yet-processed capsule + disarm `OsIndications`).
// `BootIsCompromised` is ported in `TegraUEFI.swift` (Task 8.2); `MarkGood`
// lives there too (Task 8.3, alongside `tegrauefi.go`'s own layout).
extension TegraUEFI {
    /// ESRT verdict of the last capsule attempt (`entry0` on both t234 and
    /// t264). Ports `tegrauefi.go`'s `ESRTStatusPath`.
    static let esrtStatusPath = "/sys/firmware/efi/esrt/entries/entry0/last_attempt_status"

    /// ESRT entry0 `last_attempt_status` codes (validated: t234 incident
    /// analysis + t264 Phase 1). 0 = success; 1-6 = standard UEFI capsule
    /// errors; the NVIDIA-specific codes and vendor range are documented
    /// per NVIDIA L4T. Ports `verify.go`'s constants.
    static let esrtSuccess = 0
    static let esrtUEFIErrLo = 1
    static let esrtUEFIErrHi = 6
    /// 6163: NVIDIA "CheckTheImage failed" — the capsule was rejected. NOT
    /// a cert/auth failure (test-cert capsules verify fine on a clean
    /// device); observed to be boot-chain-state dependent.
    static let esrtNvidiaCheckImageFail = 6163
    /// 6164: device SKU not in the capsule's BUP.
    static let esrtNvidiaSKUMismatch = 6164
    static let esrtNvidiaVendorLo = 0x1000
    static let esrtNvidiaVendorHi = 0x4000

    /// The `VerifyPlatformUpdate` cascade. The running rootfs's marker
    /// file is the source of truth for whether a bootloader update was
    /// part of this deployment (same rule as staging); `bootloaderUpdate`
    /// from the manifest is informational only and does not change the
    /// cascade.
    ///
    ///  1. PRIMARY:   bootloader version changed vs the value saved at swap
    ///  2. SECONDARY: ESRT `last_attempt_status == 0`
    ///  3. FALLBACK:  we booted, assume success
    ///
    /// Validated ESRT codes (t234 incident analysis + t264 Phase 1): 0
    /// success; 1-6 standard UEFI capsule errors; 6163 NVIDIA "CheckTheImage
    /// failed" / capsule rejected (boot-chain-state dependent, NOT a
    /// cert/auth failure); 6164 NVIDIA SKU mismatch; 0x1000-0x4000 NVIDIA
    /// vendor range. `nvbootctrl`'s own capsule status is NOT consulted —
    /// NVIDIA documents it as unreliable. Ports `verify.go`'s
    /// `VerifyPlatformUpdate`.
    public func verifyPlatformUpdate(bootloaderUpdate: Bool) throws {
        guard fileStore.exists(rootDir + Self.markerPath) else { return }

        // 1) version comparison.
        if let before = try? fileStore.read(blVersionBeforePath()) {
            let beforeVersion = Self.trimmed(String(decoding: before, as: UTF8.self))
            if let after = try? bootloaderVersion() {
                if beforeVersion != after {
                    try? fileStore.remove(blVersionBeforePath())
                    return  // version changed: capsule applied
                }
                // unchanged: fall through to the ESRT check.
            }
            // version unreadable: fall through to the ESRT check.
        }
        // no recorded pre-update version: fall through to the ESRT check.

        // 2) ESRT verdict.
        if let raw = try? fileStore.read(rootDir + Self.esrtStatusPath) {
            let text = Self.trimmed(String(decoding: raw, as: UTF8.self))
            guard let status = Int(text) else {
                throw TegraUEFIError.platformVerifyESRTUnparseable(text)
            }
            switch status {
            case Self.esrtSuccess:
                try? fileStore.remove(blVersionBeforePath())
                return
            case Self.esrtUEFIErrLo...Self.esrtUEFIErrHi:
                throw TegraUEFIError.platformVerifyESRTStandardError(status)
            case Self.esrtNvidiaCheckImageFail:
                throw TegraUEFIError.platformVerifyESRTCapsuleRejected
            case Self.esrtNvidiaSKUMismatch:
                throw TegraUEFIError.platformVerifyESRTSKUMismatch
            case Self.esrtNvidiaVendorLo...Self.esrtNvidiaVendorHi:
                throw TegraUEFIError.platformVerifyESRTVendorError(status)
            default:
                break  // unknown status: fall through to the boot-success fallback.
            }
        }
        // ESRT not readable: fall through to the boot-success fallback.

        // 3) fallback: the system booted to this point.
        try? fileStore.remove(blVersionBeforePath())
    }

    /// Unstages a capsule that has not been processed: removes
    /// `TEGRA_BL.Cap` from the ESP and disarms the `OsIndications` capsule
    /// bit so firmware will not look for one. No-op when nothing is
    /// staged. Ports `verify.go`'s `AbortPlatformUpdate`.
    public func abortPlatformUpdate() throws {
        var staged = false

        if let espDir = try? espMountpoint() {
            let cap = espDir + "/" + Self.espCapsuleRel
            if fileStore.exists(cap) {
                staged = true
                do {
                    try fileStore.remove(cap)
                } catch {
                    throw TegraUEFIError.abortPlatformUpdateFailed("\(error)")
                }
            }
        }

        do {
            try clearOsIndicationsCapsuleBit(osIndicationsVarPath())
        } catch {
            throw TegraUEFIError.abortPlatformUpdateFailed("\(error)")
        }

        if staged {
            try? fileStore.remove(blVersionBeforePath())
        }
    }

    /// Disarms capsule processing (rollback before reboot): clears bit 2
    /// of `OsIndications`, preserving any other bits; a missing variable
    /// is a no-op. Ports `efivar.go`'s `clearOsIndicationsCapsuleBit`.
    func clearOsIndicationsCapsuleBit(_ path: String) throws {
        guard efivarExists(path) else { return }

        let raw: [UInt8]
        do {
            raw = try EfiVar.readStatus(path)
        } catch {
            throw TegraUEFIError.osIndicationsFailed("\(error)")
        }
        guard raw.count >= 5, raw[4] & UInt8(Self.osIndicationsProcessCapsule) != 0 else {
            return  // bit not set
        }

        var value: UInt64 = 0
        for i in 0..<min(8, raw.count - 4) {
            value |= UInt64(raw[4 + i]) << (8 * i)
        }
        value &= ~Self.osIndicationsProcessCapsule

        var payload = [UInt8](repeating: 0, count: 12)
        payload[0] = 0x07
        for i in 0..<8 {
            payload[4 + i] = UInt8((value >> (8 * i)) & 0xFF)
        }
        do {
            try EfiVar.writeVar(path, payload)
        } catch {
            throw TegraUEFIError.osIndicationsFailed("\(error)")
        }
    }
}
